File: ios/BockMedia/Media/LocalPlaybackQueueResolver.swift
import Foundation

enum LocalPlaybackQueueResolver {
    static func resolve(
        repository: BockMediaRepository,
        target: PlayTarget,
        maxTracks: Int = 150
    ) async throws -> [LocalTrack] {
        let base = try await repository.resolveBaseURL()
        let offline = OfflineDownloadStore()

        func track(path: String, title: String, artist: String?, album: String?) -> LocalTrack? {
            guard let urlString = ServerURL.streamURL(base: base, filepath: path),
                  let url = URL(string: urlString) else { return nil }
            return LocalTrack(
                path: path,
                title: title,
                artist: artist,
                album: album,
                streamURL: url,
                localFileURL: offline.localFileURL(path: path)
            )
        }

        switch target {
        case .song(let path, let title):
            guard let t = track(path: path, title: title, artist: nil, album: nil) else {
                throw LocalPlaybackError.missingStreamURL
            }
            return [t]

        case .playlist(let id, let name):
            var playlistId = id
            if playlistId.isEmpty, let resolved = await repository.resolvePlaylistId(name) {
                playlistId = resolved
            }
            guard !playlistId.isEmpty else { throw LocalPlaybackError.noTracks }
            var collected: [LocalTrack] = []
            var page = 1
            while collected.count < maxTracks {
                let detail = try await repository.playlistDetail(id: playlistId, page: page, limit: 200, sortBy: "original")
                for item in detail.tracks {
                    guard let path = item.path, !path.isEmpty else { continue }
                    guard let t = track(path: path, title: item.title ?? name, artist: item.artist, album: item.album) else {
                        continue
                    }
                    collected.append(t)
                    if collected.count >= maxTracks { return collected }
                }
                if detail.tracks.isEmpty || collected.count >= detail.total { break }
                page += 1
            }
            guard !collected.isEmpty else { throw LocalPlaybackError.noTracks }
            return collected

        case .artist(let name):
            return try await loadSongPages(repository: repository, maxTracks: maxTracks) { page in
                try await repository.songs(page: page, limit: 200, artist: name)
            } map: { song, _ in
                guard let path = song.path, !path.isEmpty else { return nil }
                return track(path: path, title: song.title ?? name, artist: song.artist, album: song.album)
            }

        case .album(let name, let artist):
            return try await loadSongPages(repository: repository, maxTracks: maxTracks) { page in
                try await repository.songs(page: page, limit: 200, artist: artist, album: name)
            } map: { song, _ in
                guard let path = song.path, !path.isEmpty else { return nil }
                return track(path: path, title: song.title ?? name, artist: song.artist, album: song.album)
            }

        case .radio(_, let seedKind, let name, let path):
            switch seedKind {
            case .song:
                if let path, let t = track(path: path, title: name, artist: nil, album: nil) {
                    return [t]
                }
                fallthrough
            case .artist, .genre:
                return try await loadSongPages(repository: repository, maxTracks: min(50, maxTracks)) { page in
                    try await repository.songs(page: page, limit: 200, search: name, artist: name)
                } map: { song, _ in
                    guard let p = song.path, !p.isEmpty else { return nil }
                    return track(path: p, title: song.title ?? name, artist: song.artist, album: song.album)
                }
            }
        }
    }

    private static func loadSongPages(
        repository: BockMediaRepository,
        maxTracks: Int,
        fetch: (Int) async throws -> SongsResponse,
        map: (SongItem, Int) -> LocalTrack?
    ) async throws -> [LocalTrack] {
        var collected: [LocalTrack] = []
        var page = 1
        while collected.count < maxTracks {
            let resp = try await fetch(page)
            for song in resp.items {
                guard let t = map(song, page) else { continue }
                collected.append(t)
                if collected.count >= maxTracks { return collected }
            }
            if resp.items.isEmpty || collected.count >= resp.total { break }
            page += 1
        }
        guard !collected.isEmpty else { throw LocalPlaybackError.noTracks }
        return collected
    }
}

File: android/app/src/main/kotlin/com/bockmedia/console/media/LocalPlaybackQueueResolver.kt
package com.bockmedia.console.media

import com.bockmedia.console.data.repository.BockMediaRepository
import com.bockmedia.console.domain.model.LocalTrack
import com.bockmedia.console.domain.model.PlayTarget
import com.bockmedia.console.local.OfflineDownloadStore

class LocalPlaybackQueueResolver(
    private val repository: BockMediaRepository,
    private val offlineStore: OfflineDownloadStore,
) {
    suspend fun resolve(target: PlayTarget, shuffle: Boolean, maxTracks: Int? = null): List<LocalTrack> {
        val tracks = when (target) {
            is PlayTarget.Playlist -> loadPlaylistTracks(target.id, maxTracks)
            is PlayTarget.Artist -> loadArtistTracks(target.name, maxTracks)
            is PlayTarget.Album -> loadAlbumTracks(target.name, target.artist, maxTracks)
            is PlayTarget.Song -> listOf(songTrack(target.path, target.title, null, null))
            is PlayTarget.Radio -> when (target.seedKind) {
                PlayTarget.RadioSeedKind.Song -> {
                    target.path?.let { listOf(songTrack(it, target.name, null, null)) }
                        ?: loadArtistTracks(target.name, maxTracks)
                }
                PlayTarget.RadioSeedKind.Artist ->
                    loadArtistTracks(target.name, maxTracks)
                PlayTarget.RadioSeedKind.Genre ->
                    loadGenreRadioTracks(target, maxTracks)
            }
        }.filter { it.path.isNotBlank() }.distinctBy { it.path }

        if (tracks.isEmpty()) return emptyList()
        val capped = maxTracks?.let { tracks.take(it) } ?: tracks
        // ExoPlayer applies shuffle when enabled — do not pre-shuffle here.
        return capped
    }

    private suspend fun loadGenreRadioTracks(target: PlayTarget.Radio, maxTracks: Int?): List<LocalTrack> {
        val genre = target.genreLabel().orEmpty().ifBlank { return emptyList() }
        val cap = maxTracks ?: 150
        val libraryGenre = runCatching {
            repository.genres(limit = 200).items.let { items ->
                com.bockmedia.console.domain.model.HomeFeedRules.matchingLibraryGenreForLabel(genre, items)?.name
            }
        }.getOrNull() ?: genre
        val pool = loadGenreTracks(libraryGenre, minOf(cap, 200))
        if (pool.isEmpty()) return emptyList()
        val seed = pool.random()
        val mixed = runCatching {
            repository.resonanceRadio(seedKind = "song", path = seed.path, maxTracks = minOf(cap, 80))
        }.getOrNull()?.tracks?.mapNotNull { t ->
            val path = t.path?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
            songTrack(path, t.title ?: path, t.artist, t.album, durationSec = t.duration)
        }?.filter { it.path.isNotBlank() }
        if (!mixed.isNullOrEmpty()) return mixed
        return pool.take(cap)
    }

    private suspend fun loadPlaylistTracks(id: String, maxTracks: Int?): List<LocalTrack> {
        val collectionId = if (id.startsWith("pl-")) id else "pl-$id"
        offlineStore.readManifest(collectionId)?.let { manifest ->
            val local = offlineStore.localTracksForManifest(manifest)
            if (local.isNotEmpty()) {
                return maxTracks?.let { local.take(it) } ?: local
            }
        }
        val out = mutableListOf<LocalTrack>()
        var page = 1
        while (true) {
            val resp = repository.playlistDetail(id, page = page, limit = 200)
            resp.tracks.forEach { t ->
                val path = t.path ?: return@forEach
                out += songTrack(path, t.title ?: path, t.artist, t.album, collectionId, t.duration)
                if (maxTracks != null && out.size >= maxTracks) return out
            }
            if (out.size >= resp.total || resp.tracks.isEmpty()) break
            if (maxTracks != null && out.size >= maxTracks) break
            page++
        }
        return out
    }

    private suspend fun loadArtistTracks(artist: String, maxTracks: Int?): List<LocalTrack> {
        val out = mutableListOf<LocalTrack>()
        var page = 1
        while (true) {
            val resp = repository.songs(page = page, search = artist, artist = artist)
            resp.items.forEach { t ->
                val path = t.path ?: return@forEach
                out += songTrack(path, t.title ?: path, t.artist, t.album, durationSec = t.duration)
                if (maxTracks != null && out.size >= maxTracks) return out
            }
            if (resp.items.isEmpty() || out.size >= resp.total) break
            if (maxTracks != null && out.size >= maxTracks) break
            page++
        }
        return out
    }

    private suspend fun loadGenreTracks(genre: String, maxTracks: Int?): List<LocalTrack> {
        val out = mutableListOf<LocalTrack>()
        var page = 1
        while (true) {
            val resp = repository.songs(page = page, genre = genre, limit = 200)
            resp.items.forEach { t ->
                val path = t.path ?: return@forEach
                out += songTrack(path, t.title ?: path, t.artist, t.album, durationSec = t.duration)
                if (maxTracks != null && out.size >= maxTracks) return out
            }
            if (resp.items.isEmpty() || out.size >= resp.total) break
            if (maxTracks != null && out.size >= maxTracks) break
            page++
        }
        return out
    }

    private suspend fun loadAlbumTracks(album: String, artist: String?, maxTracks: Int?): List<LocalTrack> {
        val out = mutableListOf<LocalTrack>()
        var page = 1
        while (true) {
            val resp = repository.songs(page = page, search = album, artist = artist, album = album)
            resp.items.forEach { t ->
                val path = t.path ?: return@forEach
                out += songTrack(path, t.title ?: path, t.artist, t.album, durationSec = t.duration)
                if (maxTracks != null && out.size >= maxTracks) return out
            }
            if (resp.items.isEmpty() || out.size >= resp.total) break
            if (maxTracks != null && out.size >= maxTracks) break
            page++
        }
        return out
    }

    private suspend fun songTrack(
        path: String,
        title: String,
        artist: String?,
        album: String?,
        playlistId: String? = null,
        durationSec: Int? = null,
    ): LocalTrack {
        val local = offlineStore.localFileFor(path, playlistId)
        val durationMs = durationSec?.takeIf { it > 0 }?.times(1000L) ?: 0L
        return LocalTrack(
            path = path,
            title = title,
            artist = artist,
            album = album,
            localFile = local,
            durationMs = durationMs,
        )
    }
}
